src/main/java/com/google/devtools/build/lib/worker/WorkerParser.java

```diff
@@ -0,0 +1,144 @@
+    this.binTools = binTools;
+  }
+
+  public WorkerConfig compute(Spawn spawn, SpawnRunner.SpawnExecutionContext context)
+      throws ExecException, IOException, InterruptedException {
+    // We assume that the spawn to be executed always gets at least one @flagfile.txt or
+    // --flagfile=flagfile.txt argument, which contains the flags related to the work itself (as
+    // opposed to start-up options for the executed tool). Thus, we can extract those elements from
+    // its args and put them into the WorkRequest instead.
+    List<String> flagFiles = new ArrayList<>();
+    ImmutableList<String> workerArgs = splitSpawnArgsIntoWorkerArgsAndFlagFiles(spawn, flagFiles);
+    ImmutableMap<String, String> env =
+        localEnvProvider.rewriteLocalEnv(spawn.getEnvironment(), binTools, "/tmp");
+
+    SortedMap<PathFragment, HashCode> workerFiles =
+        WorkerFilesHash.getWorkerFilesWithHashes(
+            spawn, context.getArtifactExpander(), context.getMetadataProvider());
+
+    HashCode workerFilesCombinedHash = WorkerFilesHash.getCombinedHash(workerFiles);
+
+    WorkerKey key =
+        new WorkerKey(
+            workerArgs,
+            env,
+            execRoot,
+            Spawns.getWorkerKeyMnemonic(spawn),
+            workerFilesCombinedHash,
+            workerFiles,
+            context.speculating(),
+            multiplex && Spawns.supportsMultiplexWorkers(spawn),
+            Spawns.supportsWorkerCancellation(spawn),
+            Spawns.getWorkerProtocolFormat(spawn));
+    return new WorkerConfig(key, flagFiles);
+  }
+
+  /**
+   * Splits the command-line arguments of the {@code Spawn} into the part that is used to start the
+   * persistent worker ({@code workerArgs}) and the part that goes into the {@code WorkRequest}
+   * protobuf ({@code flagFiles}).
+   */
+  private ImmutableList<String> splitSpawnArgsIntoWorkerArgsAndFlagFiles(
+      Spawn spawn, List<String> flagFiles) throws UserExecException {
+    ImmutableList.Builder<String> workerArgs = ImmutableList.builder();
+    for (String arg : spawn.getArguments()) {
+      if (FLAG_FILE_PATTERN.matcher(arg).matches()) {
+        flagFiles.add(arg);
+      } else {
+        workerArgs.add(arg);
+      }
+    }
+
+    if (flagFiles.isEmpty()) {
+      throw createUserExecException(
+          String.format(ERROR_MESSAGE_PREFIX + REASON_NO_FLAGFILE, spawn.getMnemonic()),
+          FailureDetails.Worker.Code.NO_FLAGFILE);
+    }
+
+    ImmutableList.Builder<String> mnemonicFlags = ImmutableList.builder();
+
+    workerOptions.workerExtraFlags.stream()
+        .filter(entry -> entry.getKey().equals(spawn.getMnemonic()))
+        .forEach(entry -> mnemonicFlags.add(entry.getValue()));
+
+    return workerArgs
+        .add("--persistent_worker")
+        .addAll(MoreObjects.firstNonNull(mnemonicFlags.build(), ImmutableList.of()))
+        .build();
+  }
+
+  private static UserExecException createUserExecException(
```

Comment:
You're only using this once, no need for a method.

Reply:
Done.